### DISCUSSION TITLE
Post-Quantum TLS experiments based on KEMTLS

### DISCUSSION THREAD

**Discussion by reneme** [randombit/botan#2954] [Show and tell]
With [Kyber now merged](https://github.com/randombit/botan/pull/2872) and [the TLS 1.3 client](https://github.com/randombit/botan/pull/2922) working @hrantzsch and I combined both efforts into an experimental implementation of KEMTLS: https://github.com/neXenio/botan/pull/20.

This work is meant as a demonstrator for the work done in the [KBLS project](https://www.forschung-it-sicherheit-kommunikationssysteme.de/projekte/kbls) to showcase the [PQC algorithms added in the context of this project](https://github.com/randombit/botan/issues/2500). For now, Kyber is used as a KEM to perform a modified (KEM)TLS handshake with [an experimental server implementation](https://github.com/thomwiggers/kemtls-experiment) by @thomwiggers and @claucece. Once the addition of the Dilithium signature algorithm is finished, we might extend this demonstrator to also verify the experimental server's certificates.

Extending our TLS 1.3 implementation with the KEMTLS handshake state machine adaptions proved straight forward. Furthermore the Kyber implementation interoperated nicely. We don't expect this work to be merged upstream (hence the draft PR in the neXenio fork) but hope it might be useful or insightful for some.

Also note that there is a recent draft in the IETF TLS working group about [KEM-based Authentication for TLS 1.3](https://www.ietf.org/staging/draft-celi-wiggers-tls-authkem-01.html).

**Comment by thomwiggers** [comment]
Hi! It looks like some cool work. 

Just FYI, the AuthKEM draft is not exactly compatible with KEMTLS (for example, AuthKEM uses HPKE's API and doesn't discuss ephemeral key exchange).

**Reply by hrantzsch** [reply]
Thanks for the pointer (and all your work!) We'll keep an eye on AuthKEM and hopefully add it to Botan at one point. For this demonstrator it should be enough to be able to be compatible with your KEMTLS experimental server.

We weren't actually aware of the HPKE standardization and didn't look into AuthKEM in detail yet, so this is a valuable hint :)

**Reply by thomwiggers** [reply]
If you do get around to wanting to do AuthKEM, please reach out, we might have some more implementations specifically of authkem by then 😄

**Comment by omarama** [comment]
Hi together, 

TLS1.3 Client looks really cool.  Currently, we are working at the Dilithium signature algorithm an hopefully will finish work at the end of May. Thanks for your work @hrantzsch  & @reneme, looking forward to bring all together.

**Comment by mouse07410** [comment]
It's great to see KEMTLS implemented. Thank you! When it's finished, I'd like to see the code merged.
